nltk/metrics/distance.py

```diff
@@ -182,19 +187,125 @@ def custom_distance(file):
             labelA = frozenset([labelA])
             labelB = frozenset([labelB])
-            data[frozenset([labelA,labelB])] = float(dist)
-    return lambda x,y:data[frozenset([x,y])]
+            data[frozenset([labelA, labelB])] = float(dist)
+    return lambda x, y: data[frozenset([x, y])]
 
 
-def demo():
-    edit_distance_examples = [
-        ("rain", "shine"), ("abcdef", "acbdef"), ("language", "lnaguaeg"),
-        ("language", "lnaugage"), ("language", "lngauage")]
-    for s1, s2 in edit_distance_examples:
-        print("Edit distance between '%s' and '%s':" % (s1, s2), edit_distance(s1, s2))
-    for s1, s2 in edit_distance_examples:
-        print("Edit distance with transpositions between '%s' and '%s':" % (s1, s2), edit_distance(s1, s2, transpositions=True))
+def jaro_similarity(s1, s2):
+    """
+    Computes the Jaro similarity between 2 sequences from:
+
+        Matthew A. Jaro (1989). Advances in record linkage methodology
+        as applied to the 1985 census of Tampa Florida. Journal of the
+        American Statistical Association. 84 (406): 414-20.
+
+    The Jaro distance between is the min no. of single-character transpositions
+    required to change one word into another. The Jaro similarity formula from
+    https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance :
+
+        jaro_sim = 0 if m = 0 else 1/3 * (m/|s_1| + m/s_2 + (m-t)/m)
+
+    where:
+        - |s_i| is the length of string s_i
+        - m is the no. of matching characters
+        - t is the half no. of possible transpositions.
+    """
+    # First, store the length of the strings
+    # because they will be re-used several times.
+    len_s1, len_s2 = len(s1), len(s2)
+
+    # The upper bound of the distance for being a matched character.
+    match_bound = max(len_s1, len_s2) // 2 - 1
+
+    # Initialize the counts for matches and transpositions.
+    matches = 0  # no.of matched characters in s1 and s2
+    transpositions = 0  # no. of transpositions between s1 and s2
+    flagged_1 = []  # positions in s1 which are matches to some character in s2
+    flagged_2 = []  # positions in s2 which are matches to some character in s1
+
+    # Iterate through sequences, check for matches and compute transpositions.
+    for i in range(len_s1):     # Iterate through each character.
+        upperbound = min(i+match_bound, len_s2-1)
+        lowerbound = max(0, i-match_bound)
+        for j in range(lowerbound, upperbound+1):
+            if s1[i] == s2[j] and j not in flagged_2:
+                matches += 1
+                flagged_1.append(i)
+                flagged_2.append(j)
+                break
+    flagged_2.sort()
+    for i, j in zip(flagged_1, flagged_2):
+        if s1[i] != s2[j]:
+            transpositions += 1
+
+    if matches == 0:
+        return 0
+    else:
+        return 1/3 * (matches/len_s1 +
+                      matches/len_s2 +
+                      (matches-transpositions//2)/matches
```

Comment:
Actually, is the integer division here really necessary? Can't the `matches-transposition/2` be a float?

Reply:
No. It can't be float values as the value of t has to be an integer